docs/design/4940-reliable-loki-pipelines.md

```diff
@@ -0,0 +1,153 @@
+# Proposal: Reliable Loki pipelines
+
+* Author: Karl Persson (@kalleep)
+* Last updated: 2025-11-30
+* Discussion link: https://github.com/grafana/alloy/pull/4940
+
+## Abstract
+
+Alloy's Loki pipelines currently use channels, which limits throughput due to head-of-line blocking and can cause silent log drops during config reloads or shutdowns.
+
+This proposal introduces a function-based pipeline using a `Consumer` interface, replacing the channel-based design. Source components will call `Consume()` directly on downstream components, enabling parallel processing and returning errors that sources can use for retry logic or proper HTTP error responses.
+
+## Problem
+
+Loki pipelines in Alloy are built using (unbuffered) channels, a design inherited from promtail. 
+
+This comes with two big limitations:
+1. Throughput of each component is limited due to head-of-line blocking, where pushing to the next channel may not be possible in the presence of a slow component. An example of this is usage of [secretfilter](https://github.com/grafana/alloy/issues/3694).
+2. Because there is no way to signal back to the source, components can silently drop logs during config reload or shutdown and there is no way to detect that.
+
+Consider the following simplified config:
+```
+loki.source.file "logs" {
+  targets    = targets
+  forward_to = [loki.process.logs.receiver]
+}
+
+loki.process "logs" {
+  forward_to = [loki.write.loki.receiver]
+}
+
+loki.write "loki" {}
+```
+
+`loki.source.file` will tail all files from targets and compete to send on the channel exposed by `loki.process`. Only one entry will be processed by each stage configured in `loki.process`. If a reload happens or if Alloy is shutting down, logs could be silently dropped.
+
+## Proposal 0: Do nothing
+This architecture works in most cases, it will be hard to use slow components such as `secretfilter` because a lot of the time it's too slow.
+It's also hard to use Alloy as a gateway for loki pipelines with e.g. `loki.source.api` due to the limitations listed above.
+
+## Proposal 1: Chain function calls
```

Comment:
Could you outline for both proposal how the transition would look? 
I do think Proposal 1 looks more attractive, but I could be convinced to do something in-between Proposal 1 and 2. 

What I'm not sure about is whether we flip a switch here or make this a gradual transition. It could affect the behaviour of the pipelines, especially the error propagation. But if it is better in every way, a flip-a-switch approach is viable. If it has some trade-offs we may need to make it an opt-in, gradual transition.